session/session.go

```diff
@@ -96,13 +96,24 @@ func newZKSession(servers string, recvTimeout time.Duration, logger stdLogger, c
 	}
 
 	var event zookeeper.Event
-	select {
-	case event = <-events:
-		if event.State != zookeeper.STATE_CONNECTED {
+	for {
+		select {
+		case event = <-events:
+			switch event.State {
+				case zookeeper.STATE_AUTH_FAILED:
+					return nil, ErrZKSessionNotConnected
+				case zookeeper.STATE_EXPIRED_SESSION:
+					return nil, ErrZKSessionNotConnected
+				case zookeeper.STATE_CLOSED:
+					return nil, ErrZKSessionNotConnected
+				case zookeeper.STATE_CONNECTED:
+					break
+			}
+			break
+		case <-time.After(5 * time.Second):
 			return nil, ErrZKSessionNotConnected
 		}
-	case <-time.After(5 * time.Second):
-		return nil, ErrZKSessionNotConnected
+		break
```

Comment:
This logic still does not look right to me. You're going to break out of the `for` loop when the state is `CONNECTING` and not `CONNECTED`. For example, if the order of events was:

1) `CONNECTING`
2) `AUTH_FAILED`

With this logic, it wouldn't fail to connect because we broke out of the loop early.